packages/application-shell-connectors/src/components/application-context/application-context.tsx

```diff
@@ -0,0 +1,217 @@
+    locale: user.language,
+    timeZone: user.timeZone || defaultTimeZone,
+  };
+};
+
+// Expose only certain fields as some of them are only meant to
+// be used internally in the AppShell
+const mapProjectToApplicationContextProject = (
+  project?: TRawProject
+): TApplicationContextProject | null => {
+  if (!project) return null;
+
+  return {
+    key: project.key,
+    version: project.version,
+    name: project.name,
+    countries: project.countries,
+    currencies: project.currencies,
+    languages: project.languages,
+    ownerId: project.owner.id,
+  };
+};
+
+const createApplicationContext: <ExtendedEnvironment extends {}>(
+  environment: ExtendedEnvironment & TApplicationContextEnvironment,
+  user?: TRawUser,
+  project?: TRawProject,
+  projectDataLocale?: string
+) => TApplicationContext<ExtendedEnvironment> = (
+  environment,
+  user,
+  project,
+  projectDataLocale
+) => ({
+  environment,
+  user: mapUserToApplicationContextUser(user),
+  project: mapProjectToApplicationContextProject(project),
+  permissions: project && project.permissions ? project.permissions : null,
+  dataLocale: projectDataLocale || null,
+});
+
+const ApplicationContextProvider = <ExtendedEnvironment extends {}>(
+  props: ProviderProps<ExtendedEnvironment>
+) => (
+  <Context.Provider
+    value={createApplicationContext<ExtendedEnvironment>(
+      props.environment,
+      props.user,
+      props.project,
+      props.projectDataLocale
+    )}
+  >
+    {props.children}
+  </Context.Provider>
+);
+ApplicationContextProvider.displayName = 'ApplicationContextProvider';
+
+const ApplicationContext = <ExtendedEnvironment extends {}>(
+  props: ConsumerProps<ExtendedEnvironment>
+) => (
+  <Context.Consumer>
+    {applicationContext => props.render(applicationContext)}
+  </Context.Consumer>
+);
+ApplicationContext.displayName = 'ApplicationContext';
+
+function withApplicationContext<
+  Props extends {},
+  MappedProps extends {} = {},
+  ExtendedEnvironment extends {} = {}
+>(
+  mapApplicationContextToProps?: (
+    context: TApplicationContext<ExtendedEnvironment> | {}
+  ) => MappedProps
+) {
+  return (
+    Component: React.ComponentType<
+      | Props
+      | Props & DefaultMappedProps<ExtendedEnvironment>
+      | Props & MappedProps
+    >
+  ) => {
+    const WrappedComponent = (props: Props) => (
+      <ApplicationContext
+        render={applicationContext => {
+          const mappedProps = mapApplicationContextToProps
+            ? mapApplicationContextToProps(applicationContext)
+            : { applicationContext };
+          return <Component {...props} {...mappedProps} />;
+        }}
+      />
+    );
+    WrappedComponent.displayName = `withApplicationContext(${getDisplayName<
+      Props
+    >(Component)})`;
+    return WrappedComponent;
+  };
+}
+
+// Forward-compatibility with React Hooks 🎉
+const useApplicationContext = React.useContext
+  ? <SelectedContext extends {} = {}, ExtendedEnvironment extends {} = {}>(
```

Comment:
The hook signature is similar to the HOC, but I named the map function `selector` because it does not really inject props and simply return a value. React redux also uses those names.